### PR TITLE
Ivy actions

### LINKED ISSUE #210: Jump to field in entry buffer
I really like being able to jump around between entries in the index buffer. It would be useful too to be able to jump between fields in the entry buffer when editting entries. I sometimes have to transcribe non-machine-readable text into bibtex entries and this would make it less tedious to move around the fields. <kbd>j</kbd> is unbound in the ebib entry map, so it could be the default there too.

### PR DESCRIPTION
**Pull request by Hugo-Heagren**
This pull request adds ivy actions to ebib-jump-to-field, on keys respective of those for actions are available in the entry buffer.

Resolves #210 

I've implemented `copy-field-contents` and that was pretty straightforward. The other actions (kill, delete, yank, etc.) will be a bit more involved because I'll need to modify some of the other helper functions to take a field argument and propogate all of those changes properly. I thought I would start with a draft PR so that you could see what I've got and how I'm doing it as I go along.

***

Edit, a running list of implemented functions is useful:

- [x] ebib-copy-field-contents
- [x] ebib-kill-field-contents
- [x] ebib-yank-field-contents
- [x] ebib-delete-field-contents
- [x] ebib-edit-field
- [x] ebib-edit-multiline-field
- [x] ebib-view-field-as-help
- [x] ebib-insert-abbreviation
- [x] ebib-browse-url
- [x] ebib-browse-doi
- [x] ebib-toggle-raw
- [x] ebib-add-field

### CODE REVIEW

**Comment by joostkremers** [comment]
> I've implemented `copy-field-contents` and that was pretty straightforward. The other actions (kill, delete, yank, etc.) will be a bit more involved because I'll need to modify some of the other helper functions to take a field argument and propogate all of those changes properly.

You mean functions like `ebib--redisplay-current-field`? Yeah, that's a complication I hadn't thought about, either... 

> I thought I would start with a draft PR so that you could see what I've got and how I'm doing it as I go along.

Cool!

File: ebib-ivy.el [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+;;; Commentary:
+
+;; This file is part of Ebib, a BibTeX database manager for Emacs.  It contains
+;; the code to integrate ivy's features more tightly into the interface.
+;; To use this integration, add (require 'ebib-ivy) to init.el file.
+
+;;; Code:
+
+(require 'ebib)
```
Comment (joostkremers) [inline ebib-ivy.el:40]: I'm a bit ambivalent about this `require`. I guess it avoids a bunch of warnings in `ebib-ivy.el` and it won't do any harm, but still...

I do suspect a `(require ivy)` will be necessary, BTW.

Comment (Hugo-Heagren) [reply ebib-ivy.el:40]: > I'm a bit ambivalent about this require. I guess it avoids a bunch of warnings in ebib-ivy.el and it won't do any harm, but still...

Can I ask why? I'm pretty new to elisp and emacs packaging, so I don't know about the vagueries of this sort of thing. My reasoning was that ivy-ebib will require various (not autoloaded) functions from ebib (like `ebib-copy-field-contents`), though the reverse will never be true, so there will never be a requirement circle. I can take it out though if you think that would be better. 

> I do suspect a (require ivy) will be necessary, BTW.

Ah yes, I think so!

Comment (joostkremers) [reply ebib-ivy.el:40]: > > I'm a bit ambivalent about this require. I guess it avoids a bunch of warnings in ebib-ivy.el and it won't do any harm, but still...
> 
> Can I ask why?

Sure. :slightly_smiling_face: 

> I'm pretty new to elisp and emacs packaging, so I don't know about the vagueries of this sort of thing. My reasoning was that ivy-ebib will require various (not autoloaded) functions from ebib (like `ebib-copy-field-contents`), though the reverse will never be true, so there will never be a requirement circle.

It's really more of a philosophical thing... To me, `ebib-ivy.el` is an optional dependency of `ebib`, not the other way around, so `require` doesn't feel "right".

Though you could also argue that since `ebib.el` doesn't actually use any code from `ebib-ivy.el`, it's more of an add-on rather than a dependency, and then a case could be made that `require` is appropriate.

And you're right, it doesn't lead to any practical problems.

>  I can take it out though if you think that would be better.

Just leave it in. I may at some point decide to replace the `require` with a set of `defvar`s and/or `declace-function`s as I did in `org-ebib.el`. Or I may decide to put a `(require 'ebib)` in `org-ebib.el` and be done with it. :laughing:

**Comment by Hugo-Heagren** [comment]
Implemented killing and deleting, and it all seems to work well.

Next I'll do:
- `ebib-edit-field`
- `ebib-edit-multiline-field`
- `ebib-yank-field-contents`

Because those are in a similar vein to what I've worked on so far, should be relatively straightforward, and I can see myself using them.

***

I'm a bit of a completist though, and I hate to leave a job half done.

Of all the other commands available in `ebib-entry-mode-map`, I think these would also make sense as ivy actions:
- `ebib-view-file-in-field`
- `ebib-view-field-as-help`
- `ebib-insert-abbreviation`
- `ebib-browse-url`

I would be happy to implement these too if you think it's worth it.

I could also include `ebib-add-field`. I'm not sure what I think about this---on one hand it doesn't make sense, because the command which the actions are bound in is principally for jumping to (already existing) fields. On the other hand, similar ivy/completing-read interfaces like `find-file`, or `ivy-pass` do include functions for adding items, and I find I use them quite a lot. What do you think?

**Comment by joostkremers** [comment]
> Implemented killing and deleting, and it all seems to work well.

Great!

[...]

> I'm a bit of a completist though, and I hate to leave a job half done.
> 
> Of all the other commands available in `ebib-entry-mode-map`, I think these would also make sense as ivy actions:
> 
>     * `ebib-view-file-in-field`
> 
>     * `ebib-view-field-as-help`
> 
>     * `ebib-insert-abbreviation`
> 
>     * `ebib-browse-url`
> 
> 
> I would be happy to implement these too if you think it's worth it.

Since you're the one doing the work, you're in a better position to say if it's worth it. :slightly_smiling_face: I certainly won't reject them if you decide to add them.

> I could also include `ebib-add-field`. I'm not sure what I think about this---on one hand it doesn't make sense, because the command which the actions are bound in is principally for jumping to (already existing) fields. On the other hand, similar ivy/completing-read interfaces like `find-file`, or `ivy-pass` do include functions for adding items, and I find I use them quite a lot. What do you think?

I think it would make sense if the functionality of the entry buffer is available as much as possible as ivy actions as well, so I would be in favour of adding it.

The question is more whether you think it's worth the trouble. I can't really say how easy/difficult it would be to implement it and how much time it would take. So I'll leave the decision up to you.

**Comment by Hugo-Heagren** [comment]
> I certainly won't reject them if you decide to add them.

Awesome, I'll do those too then.

> I think it would make sense if the functionality of the entry buffer
> is available as much as possible as ivy actions as well, so I would be
> in favour of adding it.
>
> The question is more whether you think it's worth the trouble.

Well, I *imagine* (haven't looked too hard at the code yet, but
still...) that it shouldn't be too difficult to include this, so I'll
assume for now that I will. Not including it won't block any other
work either, and it isn't a priority so I guess it doesn't matter too
much right now.

***

I'm having some trouble with `ebib-yank-field-contents`. I couldn't get the documented mechanics (with the yank failing on the first try if the field already has a contents, then working thereafter) to work whatever I tried. Then I tried switching to the master branch and I realised that it just doesn't work at all for me in ebib. Does this mechanic work for you? If it's an actual bug I'll just open a new issue about it.

File: ebib.el [hunk 2]
```diff
@@ -4017,8 +4024,7 @@ With a prefix argument PFX, edit the current field directly,
 without any special treatment.  Exceptions are the \"type\" field
 and any field with a multiline value, which are always edited in
 a special manner."
-  (interactive "p")
-  (let* ((field (ebib--current-field))
+  (let* ((pfx (car current-prefix-arg))
```
Comment (joostkremers) [inline ebib.el:4027]: AFAIU `current-prefix-arg` is not always a list, so taking the `car` of it is not safe. The best method is probably to read the prefix argument in `ebib-edit-current-field` and pass it as a second optional argument to `ebib--edit-field`. Or less convoluted, use `current-prefix-arg` directly. We only want to know whether it's `nil` or not, so there's no need to examine it more closely.

File: ebib.el [hunk 3]
```diff
@@ -4073,8 +4079,27 @@ a special manner."
...
+`abstract' fields is edited as a multiline field, the `keywords'
+field adds keywords one by one, also allowing completion, and the
+\"file\" field uses filename completion and shortens filenames if
+they are in (a subdirectory of) one of the directories in
+`ebib-file-search-dirs'.
+
+With a prefix argument PFX, edit the current field directly,
+without any special treatment.  Exceptions are the \"type\" field
+and any field with a multiline value, which are always edited in
+a special manner."
```
Comment (joostkremers) [inline ebib.el:4099]: The part about the `PFX` argument is still valid, it's just that there's no argument anymore...

**Comment by joostkremers** [comment]
Looks good! Just one small comment regarding the prefix argument of `ebib-edit-field`, see the code.

**Comment by Hugo-Heagren** [comment]
Thanks, I'll amend from your comments. I think they may have got lost in the many commits, but did you see the two questions I tagged on the end of my last comment?

> I'm having some trouble with ebib-yank-field-contents. I couldn't get the documented mechanics (with the yank failing on the first try if the field already has a contents, then working thereafter) to work whatever I tried. Then I tried switching to the master branch and I realised that it just doesn't work at all for me in ebib. Does this mechanic work for you? If it's an actual bug I'll just open a new issue about it.
>
>On a similar note, what is the rationale for having init-contents as an argument in ebib-edit-multiline-field? It seems to me that all uses of it just want the initial contents to be the current contents, if any. If this is common to all uses of the function, why get the use to fetch it before the function is called, instead of just fetching it inside the function itself?

**Comment by joostkremers** [comment]
> I'm having some trouble with `ebib-yank-field-contents`. I couldn't get the documented mechanics (with the yank failing on the first try if the field already has a contents, then working thereafter) to work whatever I tried. Then I tried switching to the master branch and I realised that it just doesn't work at all for me in ebib. Does this mechanic work for you? If it's an actual bug I'll just open a new issue about it.

I guess the doc string isn't very clear... :slightly_smiling_face: What I was trying to say is that `ebib-yank-field-contents` can be invoked multiple times in a row to cycle through the kill ring, just like `C-y` followed by `M-y` does. What's not explicitly stated in the doc string is that the existing contents is only replaced if it was itself the result of a call to `ebib-yank-field-contents` immediately preceding the current call.

The manual should be clearer in this respect:

https://joostkremers.github.io/ebib/ebib-manual.html#copy-cut-kill-paste-yank-and-delete

So in short, a second yank will only replace a previous yank, not the existing contents of a field.

I updated the doc string to get rid of the confusion.

> On a similar note, what is the rationale for having `init-contents` as an argument in `ebib-edit-multiline-field`? It seems to me that all uses of it just want the initial contents to be the current contents, if any. If this is common to all uses of the function, why get the use to fetch it before the function is called, instead of just fetching it inside the function itself?

I don't recall exactly, but it was probably in analogy to `ebib--edit-normal-field`, which was added in the same commit in which `ebib-edit-multiline-field` got its arguments.

Of course, the analogy isn't perfect, because `ebib--edit-normal-field` isn't an interactive function. It would make sense to rename `ebib-edit-multiline-field` to `ebib--edit-field-as-multiline` (note the double dash) and introduce a new interactive function `ebib-edit-current-field-as-multiline`, which first gets the initial contents and then calls `ebib--edit-field-as-multiline`. `ebib--edit-field-as-multiline` could then incorporate `ebib--multiline-edit`, which is only a two-line function anyway.

If you want to make those changes, feel free to go ahead. Otherwise I'll make them some time this week.

**Comment by Hugo-Heagren** [comment]
> If you want to make those changes, feel free to go ahead.

I took a look at this locally, nothing pushed yet.

> It would make sense to rename `ebib-edit-multiline-field` to `ebib--edit-field-as-multiline` (note the double dash) and introduce a new interactive function `ebib-edit-current-field-as-multiline`, which first gets the initial contents and then calls `ebib--edit-field-as-multiline`.

This sounds sensible to me too. 

> `ebib--edit-field-as-multiline` could then incorporate `ebib--multiline-edit`, which is only a two-line function anyway.

It *could*, but `ebib--multiline-edit` is also used in `ebib-edit-preamble`. It would be easy enough to reimplement the same right behaviour in that, but since exactly the same thing *is* done twice (in `ebib--edit-field-as-multiline` and `ebib-edit-preamble`), it makes sense to me keep the function `ebib--multiline-edit`.

Final note. With all of the other functions in this PR, I've tended to just redefine them to be compatible with the ivy interface (taking only a single field name argument). In general, this has seemed pretty sensible anyway because it opens up the api, and such functions might be useful in other places. Here I'm less sure. I can't see anywhere else in the current code that a function to edit a single field as multiline would be useful. It *would* be very straightforward to write though (I've tested and it works). Shall I include it? Otherwise I would probably write it into `ebib-ivy.el` as a convenience def.

**Comment by joostkremers** [comment]
> > `ebib--edit-field-as-multiline` could then incorporate `ebib--multiline-edit`, which is only a two-line function anyway.
> 
> It _could_, but `ebib--multiline-edit` is also used in `ebib-edit-preamble`. It would be easy enough to reimplement the same right behaviour in that, but since exactly the same thing _is_ done twice (in `ebib--edit-field-as-multiline` and `ebib-edit-preamble`), it makes sense to me keep the function `ebib--multiline-edit`.

Yes, definitely. I didn't check to see if the function was used somewhere else...

> Final note. With all of the other functions in this PR, I've tended to just redefine them to be compatible with the ivy interface (taking only a single field name argument). In general, this has seemed pretty sensible anyway because it opens up the api, and such functions might be useful in other places. Here I'm less sure. I can't see anywhere else in the current code that a function to edit a single field as multiline would be useful. It _would_ be very straightforward to write though (I've tested and it works). Shall I include it? Otherwise I would probably write it into `ebib-ivy.el` as a convenience def.

Yes, please include it. I intend to implement similar actions for `embark` (and hopefully someone will contribute them for `helm`), so there will be a use for it.

**Comment by Hugo-Heagren** [comment]
Alright, done, ready for review! Pretty pleased with this overall. I'm not sure which functions you want to be named as internal api (with--double-dashes), so maybe you should have a look at the different ones I've changed/added. Apart from that I can't think of anything in particular that needs looking at.

**Comment by joostkremers** [comment]
Thanks! I haven't had a chance to look at this yet, but I'll try and find some time this weekend.

File: ebib.el (hunk 3 above)
Comment (Hugo-Heagren) [reply ebib.el:4100]: Sorted, good spot.

File: ebib.el (hunk 2 above)
Comment (Hugo-Heagren) [reply ebib.el:4028]: Sorry, just realised I hadn't dealt with this. You're right, it is a problem. We do check whether it's nil or not, but we also run `numberp` on it a bit earlier in the function. But both could be reduced to checking against nil, so I'll rewrite it that way.

Comment (Hugo-Heagren) [reply ebib.el]: > Or less convoluted, use current-prefix-arg directly

Sorted. The implementation I ended up with is almost stupidly simple, but it seems to work.

**Review by joostkremers — COMMENTED** [review]
Hi, I finally found the time to take a look at your contributions. The code looks good, I just have a few smaller comments and questions. I'd be happy to address them myself, but perhaps you'd like the git history to reflect that you addressed them. :slightly_smiling_face:

File: ebib-ivy.el [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+   ("d" ebib-delete-field-contents "delete field contents")
+   ("y" ebib-yank-field-contents "yank field contents")
+   ("e" ebib-edit-field "edit field")
+   ("m" ebib--edit-multiline-field "edit multiline field")
+   ("v" ebib-view-field-as-help "view field as help")
+   ("s" ebib-insert-abbreviation "insert abbreviation")
+   ("r" ebib-toggle-raw "toggle raw")
+   ("a" ebib-add-field "add field")
+   ("u" (lambda (_) (ebib-browse-url)) "browse url")
+   ("I" (lambda (_) (ebib-browse-doi)) "browse doi")))
```
[the remark this answers is not in the document]
Comment (Hugo-Heagren) [reply ebib-ivy.el:56]: > presumably because ivy actions need to be functions that don't take arguments?

Actually the opposite. Ivy actions have to be functions which take exactly one argument (the selected candidate). Even if we go with what you suggest below about the `arg` argument, we would still need to use something like these lambdas, because we don't want to pass the candidate as `arg` (in general, this is because `ebib-browse-url` and `ebib-browse-doi` don't take an argument to indicate which field to operate on).

To get rid of the lambdas would require defining functions (only used here) which take and ignore a single argument, then run `ebib-browse-url` / `ebib-browse-doi`. Since edge cases like this are exactly what anonymouse lambdas are for, I think we should keep them.

File: ebib.el [hunk 5]
```diff
@@ -202,21 +202,26 @@ If MARK is non-nil, `ebib-mark-face' is applied to the entry."
...
+  "Redisplay the contents of FIELD in the current buffer."
   (with-current-ebib-buffer 'entry
     ;; If the `crossref' field has changed, we need to redisplay the entire entry.
-    (let ((field (ebib--current-field)))
       (if (cl-equalp field "crossref")
           (progn
             (ebib--update-entry-buffer)
             (re-search-forward "^crossref"))
         (let ((inhibit-read-only t))
+	  (beginning-of-buffer)
```
[the remark this answers is not in the document]
Comment (Hugo-Heagren) [reply ebib.el]: Good to know, sorted.

File: ebib.el [hunk 6]
```diff
@@ -3796,14 +3802,16 @@ was called interactively."
...
-    (goto-char location)
-    (beginning-of-line)))
+                         (re-search-forward (concat "^" field) nil t))))
+    (if location
+	(goto-char location)
+      (beginning-of-line))))
 
-(defun ebib-add-field (field)
+(defun ebib-add-field ()
   "Add FIELD to the current entry."
```
[the remark this answers is not in the document]
Comment (Hugo-Heagren) [reply ebib.el:3811]: So as above, ivy actions actually need to take exactly one argument. That alone makes this a bug, because the function *doesn't* take any arguments.

The *reason* it doesn't take any arguments is that I had thought calling a function from an ivy action was a non-interactive use so I needed to make sure that the function prompted for a string whatever the commit message was (I'm really glad I wrote a good commit message for that one!).

In fact, using the `(interactive "s...")` as before *does* work (presumably ivy accounts for this somehow). What happens is that the current input is used as the `field` arg. This makes sense workflow-wise: if you try to jump to a nonexistent field, you probably want to add that field. However, there once you hit <kbd>a</kbd> to add a new field, there is no opportunity to amend or add the input for the field name, you are just prompted for a value. I don't think this is quite what we want. For me it's most natural to use ivy's regex matching to see that a certain field isn't present, *then* do the actoin to add. I would expect to be prompted, perhaps with my previous input as a defualt in the prompt? What do you think? (Not that it has anything to do with ivy, but I wonder if it might be useful to have add-field work a bit like the keyword prompt, allowing new additions, but also selecting from a list of everything *already* added in the database? Just a thought.)

File: ebib.el [hunk 7]
```diff
@@ -4122,27 +4146,30 @@ viewed."
         (while (ebib--line-at-point)
           (forward-line direction))))))
 
-(defun ebib-copy-field-contents ()
-  "Copy the contents of the current field to the kill ring.
+(defun ebib-copy-field-contents (field)
+  "Copy the contents of named FIELD to the kill ring.
```
[the remark this answers is not in the document]
Comment (Hugo-Heagren) [reply ebib.el]: Done.

**Comment by joostkremers** [comment]
> > Hi, apologies for the long delay.
> 
> No worries, we all have busy lives!

I' starting a new job Monday and I actually had some free time, but it's amazing how quickly that fills up when you have a family. :smile: 

> Yes, that's right. There are two things with `ebib-add-field`. The first is the buggy behaviour in the ivy action, and making it prompt for a field name with previous input as default. I'd like to solve that in this PR because it has to do with the ivy actions. Hopefully I'll get a chance to look at it over the weekend. Then there's my separate suggestion that `ebib-add-field` could complete on names of all fields already used in the database. I think that's worth doing, but it can wait for a separate PR another day.

OK, I'll wait with merging until you find time to fix the first issue and I'll look forward to the PR for the second issue. :slightly_smiling_face: 

> Magit to the rescue, I'm sure. Don't worry about it, I'm at that stage in learning development where problems like this are still interesting to solve, so I'll have some fun learning about what the correct git way to solve this would be :).

:relieved: I think the recipe I gave (reset the branch to the last commit you made, then rebase it on current master so I can push the relevant commits again) should get you going. It's what I would try, anyway.

**Comment by Hugo-Heagren** [comment]
Right, sorted, I think. I:

- Fixed my local branch and updated everything accordingly.
- Pulled (rebasing local commits) from master, such that the local branch ivy-actions rebased over c11ffdf and b829aac.
- Force-pushed to my own fork remote, such that it also now includes c11ffdf and b829aac, and it's latest commit is by me (bfd5716).

So I think if you push your own further commits to the fork, then merge it (or vice versa) we should be good!

**Comment by joostkremers** [comment]
> Right, sorted, I think. I:
[...] 
> So I think if you push your own further commits to the fork, then merge it (or vice versa) we should be good!

Yes, I just did and it looks good. Thanks!

Note to self: next time, check more carefully what you're pushing... :grimacing: 

> (It's worth admitting that the solution I found for prompting with current input is not perfect. It's a bit of a kludge, and if the input is close enough to a candidate, it will prompt with the candidate instead.

You mean because `read-string`'s `INITIAL-INPUT` is immediately matched upon by ivy? Would it be possible to use the `DEFAULT-VALUE` argument instead and add a remark to the prompt informing the user what happens when they just hit `RET`? I don't know if that works with ivy, though. You may be able to use `ebib--completion-finish-key`.

**Comment by Hugo-Heagren** [comment]
> You mean because read-string's INITIAL-INPUT is immediately matched upon by ivy?

Not quite. `read-string` doesn't take a list, so Ivy doesn't do any matching. Rather, I couldn't (easily) find a reliable way of finding and capturing the ivy minibuffer input to *before* using `read-sting`, to put into the `INITIAL-INPUT` later on. The hack is to appeal to `ivy-last`, a struct representing the last ivy session, and extract the `current` slot. This is the current candidate if there is one, or the current input if not, so it works in the most important cases, where what you've input doesn't match any candidate.

**Comment by joostkremers** [comment]
> > You mean because read-string's INITIAL-INPUT is immediately matched upon by ivy?
> 
> Not quite. `read-string` doesn't take a list, so Ivy doesn't do any matching.

Right, `read-string` doesn't do completion.

[snip explanation]

I don't know ivy's internals, so I don't really follow... But IIUC your "hack" isn't in the current PR, right?

**Comment by Hugo-Heagren** [comment]
> But IIUC your "hack" isn't in the current PR, right?

TL;DR: no it isn't, but the behaviour isn't perfect, but it would take quite a lot of work to *make* it perfect, so the PR is ready.

Actually, sorry, I slightly misrepresented things. What happened was that I tried a couple of different ways and came up with the one above. I then realised that this is the same as Ivy's basic behaviour (return the current candidate if there is one, return minibuffer input if not), so I didn't need the 'hack' code after all, I could still get the same behaviour by just using Ivy's standard approach. This is why the ivy code in the actual PR looks so simple, but doesn't quite do exactly what I wanted it to do. However, it would be a lot of work to *make* it do that, so I think for now it's fine.

**Comment by joostkremers** [comment]
I just merged this PR into master and pushed it to Github. Github doesn't seem to notice that it's been merged, probably because of the fact that I did a fast-forward merge, but it is there.

Thank you very much for providing this functionality!
